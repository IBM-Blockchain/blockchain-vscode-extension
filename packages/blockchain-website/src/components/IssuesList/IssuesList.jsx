import React, { Component } from 'react';
import './IssuesList.scss'; 
import 'bootstrap/scss/bootstrap.scss';
import 'font-awesome/scss/font-awesome.scss';


let githubProject = 'IBM-Blockchain/blockchain-vscode-extension';

//Github URLs
let issuesHtmlUrl = 'https://github.com/'+githubProject+'/issues?q=';
let newIssueUrl = 'https://github.com/'+githubProject+'/issues/new';

class IssuesList extends Component {
    getLabel(issue) {
        if (issue.milestone) {
            let labelJSX;

            let issueLabels = issue.labels;
            let bug = false;
            issueLabels.map((label) => {
                if (label.name === 'bug') {
                    bug = true;
                }
            });

            if (bug) {
                labelJSX = (
                <h3 className='label bg-bug'> {'bug'} 
                </h3>
                )
            } else {
                labelJSX = (
                <h3 className='label bg-feature'> {'feature'} 
                </h3>
                )
            }
            return labelJSX

        } else {
            return ''
        }
    }

    render(){
        let issueItems = (
            <p className='alert alert-info' key='info'>Please wait, loading status information</p>
        );
        if (this.props.error) {
            issueItems = (
                <p className='alert alert-danger' key='error'>{this.props.error}</p>
            );
        } else if (this.props.issues !== undefined) {
            issueItems = this.props.issues.map((issue) => {
                let creationDate = new Date(issue.created_at);
                let updateDate = new Date(issue.updated_at);
                let updateOrCloseDate = 'Updated: '+ updateDate.toLocaleDateString();
                let className = "panel-danger";
                if (issue.state === 'closed') {
                    className = 'panel-success';
                    updateOrCloseDate = 'Resolved: ' + updateDate.toLocaleDateString();
                }

                return (
                    <div className={"panel " + className}>
                        <div className="panel-heading">
                            <h3 className='panel-title'>
                                {(issue.state === 'closed' ? 'RESOLVED: ' : '') + issue.title + ' (#' + issue.number + ')'}
                            </h3>
                            {this.getLabel(issue)}
                        </div>
                        <div className='panel-body'>
                            <p>
                                <span>{'Reported: ' + creationDate.toLocaleDateString()}</span>
                                <span className='float-right'>{updateOrCloseDate}</span>
                            </p>
                            <hr className='hr'></hr>
                            <p>
                                <a href={issue.html_url}>{'View on GitHub (' + issue.comments + ' comments so far)'}</a>
                            </p>
                        </div>
                    </div>
                );
            });
            if (issueItems.length === 0) {
                if (this.props.newestMilestone) {
                    issueItems = (
                        <p className='alert alert-danger'>
                            {'No completed issues!'}
                        </p>
                    )
                } else {
                    issueItems = (
                        <p className='alert alert-success'>
                            {'No incidents found!'}
                        </p>
                    )
                }
            }
        }

        let milestoneUrl = 'https://github.com/'+githubProject+'/milestone/';
        milestoneUrl += this.props.milestoneNumber;
        const seeMilestoneLink = (this.props.newestMilestone !== undefined) ? (<a href={milestoneUrl}> {'See current milestone'} </a>) : (<a href={issuesHtmlUrl}> {'See all on GitHub'}</a>);
        const reportIncident = (this.props.newestMilestone !== undefined) ? ('') : (<a href={newIssueUrl}>{'Report an incident'}</a>);


        return(
            <div>
                <h2>
                    {this.props.title}
                </h2>
                <p>
                    {seeMilestoneLink}
                    {this.props.newestMilestone !== undefined ? ('') : (' // ')}
                    {reportIncident}
                </p>
                {issueItems}
            </div>
        )
    }
}

export default IssuesList;