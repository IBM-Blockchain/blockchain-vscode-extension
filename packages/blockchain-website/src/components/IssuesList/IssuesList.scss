.flex-container {
    display: flex;
    flex-flow: row wrap;
}

.panel {
    background-color: white;
    width: 309px;
    margin: 1px;
    position: relative;
    height: 165px;
}

.no-issues {
    background-image: linear-gradient(180deg, #FFFFFF 0%, #F3F3F3 100%);
    width: 1250px;
    height: 165px;
    padding: 16px;
    line-height: 16px;
    margin: 1px; 
}

.issue-title {
    margin-left: 16px;
    margin-right: 21px;
    margin-bottom: 7px;
    font-size: 14px;
    color: #161616;
    letter-spacing: 0.16px;
    line-height: 18px;
}

.div-title {
    display: flex;
    flex-flow: row wrap;
    padding-top: 20px;
    max-width: 1244px;
    font-size: 16px;
    color: #161616;
    letter-spacing: 0;
    line-height: 22px;
    padding-bottom: 5px; 
}

.date {
    margin-left: 16px;
    margin-right: 21px;
    font-size: 12px;
    color: #525252;
    letter-spacing: 0.32px;
    text-align: left;
    line-height: 16px;
}

.link {
    text-decoration: none;
    font-size: 14px;
    color: #0F62FE;
    letter-spacing: 0.16px;
    line-height: 18px;
}

.view-issue {
    position: absolute;
    bottom: 16px;
    right: 16px;
}

.button-container {
    margin-left: auto;
    max-width: 450px;
    max-height: 30px;
    @media (max-width: 620px) { // If screen width small, move buttons directly below title (pin to the left of the container)
        margin-right: auto;
    }
}

.button {
    text-decoration: none;
    display: inline-block;
    background: url('../../add.svg');
    background-repeat: no-repeat;
    background-size: 16px;
    background-color: #0f62fe;
    background-position-x: 169px;
    background-position-y: 8px;
    height: 19px;
    width: 185px;
    color: white;
    font-size: 14px;
    padding-top: 3px;
    padding-bottom: 7px;
    padding-left: 16px;
}

.new-issue-position {
    margin-left: 2rem;
}

.view-all-position {
    margin-left: auto;
}

.view-milestone {
    margin-left: auto;
}

.label {
    border-radius: 4px;
    display: inline;
    background-color: #b5b5b5;
    color: #fff;
    font-size: 12px;
    font-size: 0.75rem;
    line-height: 2;
    font-weight: bold;
    padding: 3px 7px;
    text-align: center;
    vertical-align: baseline;
    white-space: nowrap;
}

.bg-sev1 {
    background-color: #da1e28 !important;
}
  
.bg-sev2 {
    background-color: #ff832b !important;
}

.bg-sev3 {
    background-color: #f1c21b !important;
}

.label-div {
    margin-left: 16px;
    margin-right: 21px;
}