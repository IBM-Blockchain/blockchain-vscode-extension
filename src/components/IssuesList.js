import React, { Component } from 'react';
import '../App.css'; 
import 'bootstrap/dist/css/bootstrap.min.css';
import 'font-awesome/css/font-awesome.min.css';


let githubProject = 'IBM-Blockchain/blockchain-vscode-extension';

//Github URLs
let issuesHtmlUrl = 'https://github.com/'+githubProject+'/issues?q=';
let newIssueUrl = 'https://github.com/'+githubProject+'/issues/new';

const markDownParser = function(str){
    var lines = str.split("\n");
    var output = [];
    for(var i=0;i<lines.length;i++){
        var line=lines[i];
        var titleRegex = /^(\#+)\s+(.*)/ig;
        var result = titleRegex.exec(line)
        if (result){
            var order = result[1].length;
            output.push(
            <h key={i}>
                {(order+2).toString()}
                {result[2]}
            </h>
            )
            continue;
        }
        output.push(
            <p key={i}>
                {line}
            </p>
        )
    }
    return output;
}

class IssuesList extends Component {
    getLabel(issue) {
        if (issue.milestone) {
            let labelJSX;

            let issueLabels = issue.labels;
            let bug = false;
            issueLabels.map((label) => {
                if (label.name === 'bug') {
                    bug = true;
                }
            });

            if (bug) {
                labelJSX = (
                <h3 className='label bg-bug'> {'bug'} 
                </h3>
                )
            } else {
                labelJSX = (
                <h3 className='label bg-feature'> {'feature'} 
                </h3>
                )
            }
            return labelJSX

        } else {
            return ''
        }
    }

    render(){
        let issueItems = (
            <p className='alert alert-info' key='info'>Please wait, loading status information</p>
        );
        if (this.props.error) {
            issueItems = (
                <p className='alert alert-danger' key='error'>{this.props.error}</p>
            );
        } else if (this.props.issues !== undefined) {
            issueItems = this.props.issues.map((issue) => {
                if (this.props.newestMilestone) {
                    let found = false;

                    if (issue.milestone) {
                        if (issue.milestone.title === this.props.newestMilestone.title)
                        found = true;
                    }

                    if (!found) {
                        return;
                    }
                }

                let creationDate = new Date(issue.created_at);
                let updateDate = new Date(issue.updated_at);
                let updateOrCloseDate = 'Updated: '+ updateDate.toLocaleDateString();
                let className = "panel-danger";
                if (issue.state === 'closed') {
                    className = 'panel-success';
                    updateOrCloseDate = 'Resolved: ' + updateDate.toLocaleDateString();
                }

                return (
                    <div className={"panel " + className}>
                        <div className="panel-heading">
                            <h3 className='panel-title'>
                                {(issue.state === 'closed' ? 'RESOLVED: ' : '') + issue.title + ' (#' + issue.number + ')'}
                            </h3>
                            {this.getLabel(issue)}
                        </div>
                        <div className='panel-body'>
                            <p>
                                <span>{'Reported: ' + creationDate.toLocaleDateString()}</span>
                                <span className='float-right'>{updateOrCloseDate}</span>
                            </p>
                            <hr className='hr'></hr>
                            {issue.state === 'open' ? markDownParser(issue.body): ''}
                            <p>
                                <a href={issue.html_url}>{'View on GitHub (' + issue.comments + ' comments so far)'}</a>
                            </p>
                        </div>
                    </div>
                );
            });
            if (issueItems.length === 0) {
                issueItems = (
                    <p className='alert alert-success'>
                        {'No incidents found!'}
                    </p>
                )
            }
        }

        let spinnerClass = 'fa fa-refresh';
        if (this.props.refreshing) {
            spinnerClass += ' fa-spin';
        }

        let milestoneUrl = 'https://github.com/'+githubProject+'/milestone/';
        milestoneUrl += this.props.milestoneNumber;
        const seeMilestoneLink = (this.props.newestMilestone !== undefined) ? (<a href={milestoneUrl}> {'See current milestone'} </a>) : (<a href={issuesHtmlUrl}> {'See all on GitHub'}</a>);

        return(
            <div>
                <h2>
                    {this.props.title}
                    <a className='float-right' href='#'>
                        <i className={spinnerClass} onClick={this.props.refreshIssues}></i>
                    </a>
                </h2>
                <p>
                    {seeMilestoneLink}
                    {' // '}
                    <a href={newIssueUrl}>{'Report an incident'}</a>
                </p>
                {issueItems}
            </div>
        )
    }
}

export default IssuesList;