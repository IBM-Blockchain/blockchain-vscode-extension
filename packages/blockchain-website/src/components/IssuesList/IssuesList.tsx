import React, { Component } from 'react';
import './IssuesList.scss';

const githubProject: string = 'IBM-Blockchain/blockchain-vscode-extension';

// Github URLs
const issuesHtmlUrl: string = 'https://github.com/' + githubProject + '/issues?q=';
const newIssueUrl: string = 'https://github.com/' + githubProject + '/issues/new';

interface IssueProps {
    issues?: any;
    error?: string;
    title?: string;
    issueLabel?: string;
    newestMilestone?: any;
    milestoneNumber?: string;
}

class IssuesList extends Component<IssueProps, {}> {

    getLabel(issue: any): JSX.Element {
        if (!issue.milestone) {
            let labelJSX: JSX.Element = <></>;

            const issueLabels: [{id: number, node_id: string, url: string, name: string, color: string, default: boolean, description: string}] = issue.labels;
            issueLabels.map((label: {id: number, node_id: string, url: string, name: string, color: string, default: boolean, description: string}) => {
                if (label.name === 'sev1') {
                    labelJSX = (
                        <h3 className='label bg-sev1'> {'sev1'}
                        </h3>
                    );
                } else if (label.name === 'sev2') {
                    labelJSX = (
                        <h3 className='label bg-sev2'> {'sev2'}
                        </h3>
                    );
                } else {
                    labelJSX = (
                        <h3 className='label bg-sev3'> {'sev3'}
                        </h3>
                    );
                }
            });
            return labelJSX;

        } else {
            return <></>;
        }
    }

    render(): JSX.Element {
        let issueItems: any = (
            <div className='no-issues'>Please wait, loading status information</div>
        );
        if (this.props.error) {
            issueItems = (
                <div className='no-issues'> {this.props.error}</div>
            );
        } else if (this.props.issues !== undefined) {
            // Sort the incidents by severity
            let issuesArray: any = this.props.issues;
            const sortedArray: any = [];
            if (!this.props.newestMilestone) {
                issuesArray.map((issue: any) => {
                    let labels: any = issue.labels;
                    labels = labels.filter((label: {id: number, node_id: string, url: string, name: string, color: string, default: boolean, description: string}) => {
                        return(label.name.startsWith('sev'));
                    });
                    issue.labels = labels;
                    sortedArray.push(issue);
                });
                issuesArray = sortedArray.sort((a: any, b: any) => a.labels[0].name.localeCompare(b.labels[0].name));
            }

            issueItems = issuesArray.map((issue: any) => {
                const creationDate: Date = new Date(issue.created_at);
                const updateDate: Date = new Date(issue.updated_at);
                let updateOrCloseDate: string = 'Updated: ' + updateDate.toLocaleDateString();
                if (issue.state === 'closed') {
                    updateOrCloseDate = 'Resolved: ' + updateDate.toLocaleDateString();
                }

                return (
                    <div className='panel'>
                        <div>
                            <p className='issue-title'>
                                {(issue.state === 'closed' ? 'RESOLVED: ' : '') + issue.title + ' (#' + issue.number + ')'}
                            </p>
                        </div>
                        <div className='label-div'>
                            {this.getLabel(issue)}
                        </div>
                        <div>
                            <span className='date'>{'Reported: ' + creationDate.toLocaleDateString()}</span>
                            <br></br>
                            <span className='date'>{updateOrCloseDate}</span>
                        </div>
                        <a className='link view-issue' href={issue.html_url}>{'View on GitHub'}</a>
                    </div>
                );
            });

            if (issueItems.length === 0) {
                if (this.props.newestMilestone) { // Only issues in current milestoe have this property.
                    issueItems = (
                        <div className='no-issues'>
                            No completed fixes yet!
                        </div>
                    );
                } else {
                    issueItems = (
                        <div className='no-issues'>
                            No incidents found!
                        </div>
                    );
                }
            }
        }

        let milestoneUrl: string = 'https://github.com/' + githubProject + '/milestone/';
        milestoneUrl += this.props.milestoneNumber;
        const milestoneRelease: boolean = (this.props.newestMilestone !== undefined) ? true : false;
        let reportIncident: JSX.Element;

        if (this.props.newestMilestone) {
            reportIncident = (<></>);
        } else {
            reportIncident = <span className='new-issue-position'>
                <a className='button' href={newIssueUrl}>Report a new issue</a>
                </span>;
        }

        return(
            <div>
                <div className='div-title'>
                    {this.props.title}

                    <div className='button-container'>
                        {milestoneRelease ? (
                            <span className='view-milestone'>
                                <a className='link' href={milestoneUrl}> {'See current milestone'} </a>
                            </span>
                        ) : (
                            <span className='view-all-position'>
                                <a className='link' href={issuesHtmlUrl}> {'View all on GitHub'} </a>
                            </span>
                        )}
                        {reportIncident}
                    </div>
                </div>
                <div className='flex-container'>
                    {issueItems}
                </div>
            </div>
        );
    }
}

export default IssuesList;
