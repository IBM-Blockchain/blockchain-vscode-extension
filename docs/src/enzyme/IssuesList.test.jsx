import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
import IssuesList from '../components/IssuesList/IssuesList.jsx';
chai.should();
chai.use(sinonChai);

describe('IssuesList', () => {
    let mySandBox;

    beforeEach(async() => {
        mySandBox = sinon.createSandbox();
    });

    afterEach(async () => {
        mySandBox.restore();
    })

    it('should render the expected snapshot', async() => {
        const component = renderer
            .create(<IssuesList />)
            .toJSON(); 
        expect(component).toMatchSnapshot();
    });
    
    it('should show error test if error is passed as a prop', async() => {
        const component = await mount(<IssuesList error={'someError'}/>);
        component.text().includes('someError').should.be.true;
    });

    it('should show message if no incidents to show', async() => {
        const component = await mount(<IssuesList issues={[]}/>);
        component.text().includes('No incidents found!').should.be.true;
    });

    it('should show message if no completed issues to show from the most recent milestone', async() => {
        const component = await mount(<IssuesList issues={[]} newestMilestone={'someMilestone'}/>);
        component.text().includes('No completed issues!').should.be.true;
    });

    it('should show any incidents', async() => {
        const incidents = {
            url: 'someURL',
            repositoryurl: 'someOtherURL',
            number: 1946,
            title: 'someIncident issue',
            state: 'open'
        }
        const component = await mount(<IssuesList issues={[incidents]}/>);
        component.text().includes('someIncident issue').should.be.true;
    });

    it('should show any completed issues for the next release', async() => {
        const completedIssues = [{
            url: 'someURL',
            repositoryurl: 'someOtherURL',
            number: 1932,
            title: 'Some completed issue',
            state: 'closed',
            labels: [{
                id : 'someLabelID',
                name: 'bug',
                description: 'someLabelDescription'
            }],
            milestone: {
                number: 'milestoneNumber',
                id: "someMilestoneID"
            }
        },
        {
            url: 'someURL1',
            repositoryurl: 'someOtherURL1',
            number: 1922,
            title: 'Some other completed issue',
            state: 'closed',
            labels: [{
                id : 'someOtherLabelID',
                name: 'enhancement',
                description: 'someOtherLabelDescription'
            }],
            milestone: {
                number: 'milestoneNumber',
                id: "someMilestoneID"
            }
        }]
        const component = await mount(<IssuesList issues={completedIssues} newestMilestone={'someMilestone'}/>);
        component.text().includes('Some completed issue').should.be.true;
        component.text().includes('Some other completed issue').should.be.true;
        component.text().includes('bug').should.be.true;
        component.text().includes('feature').should.be.true;
    });
});