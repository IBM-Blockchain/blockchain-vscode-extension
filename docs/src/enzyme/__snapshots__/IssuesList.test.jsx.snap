// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesList should render the expected snapshot 1`] = `
<div>
  <h2>
    <a
      className="float-right"
      href="#"
    >
      <i>
         
        <svg
          aria-hidden={true}
          focusable="false"
          height={24}
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "willChange": "transform",
            }
          }
          viewBox="0 0 32 32"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 10H6.78A11 11 0 0127 16h2A13 13 0 006 7.68V4H4v8h8zM20 22h5.22A11 11 0 015 16H3a13 13 0 0023 8.32V28h2V20H20z"
          />
        </svg>
         
      </i>
    </a>
  </h2>
  <p>
    <a
      href="https://github.com/IBM-Blockchain/blockchain-vscode-extension/issues?q="
    >
       
      See all on GitHub
    </a>
     // 
    <a
      href="https://github.com/IBM-Blockchain/blockchain-vscode-extension/issues/new"
    >
      Report an incident
    </a>
  </p>
  <p
    className="alert alert-info"
  >
    Please wait, loading status information
  </p>
</div>
`;
